### PR TITLE
Add http auth scheme

### PR DESCRIPTION
**Pull request by KhraksMamtsov**
Add Basic and Bearer Authentication
https://swagger.io/docs/specification/authentication/bearer-authentication/
https://swagger.io/docs/specification/authentication/basic-authentication/

add one more parameter after options in endpoint constructor
handle it in openapi, on the server side and on the client side and in types

### CODE REVIEW

**Comment by KhraksMamtsov** [comment]
Don't understand how to fix build and pages steps(
I have added doc-annotations

**Comment by sukovanej** [comment]
it reports `Missing documentation in src/SecurityScheme.ts module` so you're probably missing a docstring somewhere

**Comment by sukovanej** [comment]
Hey, I'd prefer to let the structure of tsconfigs remain as is. The current state is how I'm used to it and also how other effect packages do it. Based on the commit message, if it's a Webstorm problem it should be probably addressed in a local configuration instead. BTW what are the problems you're trying to fix?

**Comment by KhraksMamtsov** [comment]
`pnpm run check` - shows 181 errors( I don't understand why
sources of errors in the `eamples` and `test` folders
Maybe because of that build step fails?

**Comment by sukovanej** [comment]
I see, run `pnpm build` locally. It will regenerate the `index.ts`.

**Comment by KhraksMamtsov** [comment]
@sukovanej what do you think? Should I add more tests or clarify something somehow?

**Comment by sukovanej** [comment]
Hey, thanks a lot for the PR. I'll go through it during the weekend.

**Comment by KhraksMamtsov** [comment]
@sukovanej sorry, but can I get some feedback from you?

**Comment by sukovanej** [comment]
Hey, sorry, I didn't get to it. It's on the todo list.

**Comment by sukovanej** [comment]
Hey, finally went through it. 

Regarding the proposed API, I'm worried the `Api[method](...)` parameters are really starting bloating. I'd consider making the security a an optional part of the `EndpointSchemas` (can be renamed to reflect the purpose better). Also, I started thinking about taking the pipeable way. Something like this

```ts
cosnt api = Api.api().pipe(
  Api.get('operationName', '/operation').pipe(
    Api.responseBody(MySchema),
    Api.requestSearchParams(MySearchParamsSchema),
    Api.security(Security.basic(...)),
    ...
  ),
  ...
)
```

Of course, this is absolutely out of the scope of this MR. I'm just thinking out loud how to resolve the parameters bloating.

Another thing that comes to my mind is that usually bigger portion of an API is being secured by an auth (usually, all but some health or ping endpoints are being secured in case of non-internal APIs). So it would be nice to come up with a way to make this work on all `api`, `group` and `endpoint` level. With the pipeable API, we could potentially go the usual way and make this api middleware-like, meaning exposing a combinator imposing the security for the whole api data structure passed to it.

Another question is the handling part of the security. I understand, you are proposing solely an extension to enable a structural validation and OpenApi generation. The actual authentication / authorization is left to the user-land code to deal with. That is okay for now, but we should make sure it is documented somewhere because a lib user might expect there is a designed way to do it.

Anyway, I'm probably thinking too much out of scope here. The scope here is to do the validation per endpoint and make the info available in the OpenApi. So, I'd only make the security part of schemas or options parameters. Implementation wise, please add an internal module for the SecurityScheme module and put only public API symbols there. I think I noticed some typos, `x`-named variables, etc, usual stuff to be cleaned up. Let's merge it after these changes and release together with the new effect minor release changes.

Thanks again for the PR!

**Comment by KhraksMamtsov** [comment]
Honestly, as for your thoughts on the future development of the library, I think the same way. (parameters explosion, grouping security, and so on)
Yes, I will make the necessary edits.
As for the joint release, please do not wait for my edits. Let's do it regardless.

**Comment by KhraksMamtsov** [comment]
@sukovanej 
- add some doc
- move `security` parameter into `options` of endpoint
- change inferred token type in case single security scheme specified
- rename `x` variables
- add internal `security-scheme` module

**Review by sukovanej — CHANGES_REQUESTED** [review]

File: tsconfig.test.json
[the remark this answers is not in the document]
Comment (KhraksMamtsov) [reply tsconfig.test.json:1]: You are right
I have removed them

**Pull request merged**